File: backend/pdf_generator.py
from fpdf import FPDF

def generate_invoice_pdf(order_id, personal_info, delivery_method, payment_method, shopping_bag, total_price, logo_path="logo.png", as_bytes=False):
    pdf = FPDF()
    pdf.add_page()

    # Header
    pdf.set_font("Arial", style="B", size=16)
    pdf.cell(200, 10, txt="Melangé Invoice", ln=True, align="C")
    pdf.ln(5)

    # Add Logo
    if logo_path:
        pdf.image(logo_path, x=90, y=25, w=30)  # Smaller logo with width 30mm, centered
        pdf.ln(35)  # Add vertical spacing to move content below the logo

    # Order ID
    pdf.set_font("Arial", size=12)
    pdf.cell(200, 10, txt=f"Order ID: {order_id}", ln=True, align="C")
    pdf.ln(10)

    # Personal Information
    pdf.set_font("Arial", style="B", size=12)
    pdf.cell(0, 10, txt="Personal Information", ln=True, border="B")
    pdf.set_font("Arial", size=10)
    pdf.cell(0, 10, txt=f"Name: {personal_info['name']}", ln=True)
    pdf.cell(0, 10, txt=f"Surname: {personal_info['surname']}", ln=True)
    pdf.cell(0, 10, txt=f"Address: {personal_info['address']}", ln=True)
    pdf.cell(0, 10, txt=f"Email: {personal_info['email']}", ln=True)
    pdf.ln(10)

    # Delivery Method
    pdf.set_font("Arial", style="B", size=12)
    pdf.cell(0, 10, txt="Delivery Method", ln=True, border="B")
    pdf.set_font("Arial", size=10)
    pdf.cell(0, 10, txt=delivery_method.capitalize(), ln=True)
    pdf.ln(10)

    # Payment Information
    pdf.set_font("Arial", style="B", size=12)
    pdf.cell(0, 10, txt="Payment Information", ln=True, border="B")
    pdf.set_font("Arial", size=10)
    if payment_method.get("type") == "credit":
        pdf.cell(0, 10, txt=f"Card Holder: {payment_method['cardHolderName']}", ln=True)
        pdf.cell(0, 10, txt=f"Card Number: **** **** **** {payment_method['cardNumber'][-4:]}", ln=True)
    pdf.ln(10)

    # Shopping Bag (Table)
    pdf.set_font("Arial", style="B", size=12)
    pdf.cell(0, 10, txt="Shopping Bag", ln=True, border="B")
    pdf.set_font("Arial", size=10)

    # Table Header
    pdf.cell(80, 10, txt="Product Name", border=1, align="C")
    pdf.cell(30, 10, txt="Color", border=1, align="C")
    pdf.cell(20, 10, txt="Size", border=1, align="C")
    pdf.cell(30, 10, txt="Quantity", border=1, align="C")
    pdf.cell(30, 10, txt="Price", border=1, align="C")
    pdf.ln()

    # Table Rows
    for item in shopping_bag:
        pdf.cell(80, 10, txt=item["product_name"], border=1)
        pdf.cell(30, 10, txt=item["color_name"], border=1, align="C")
        pdf.cell(20, 10, txt=item["size_name"], border=1, align="C")
        pdf.cell(30, 10, txt=str(item["quantity"]), border=1, align="C")
        pdf.cell(30, 10, txt=f"${item['price']:.2f}", border=1, align="C")
        pdf.ln()

    # Total Price
    pdf.ln(10)
    pdf.set_font("Arial", style="B", size=12)
    pdf.cell(0, 10, txt=f"Total Price: ${total_price:.2f}", ln=True, align="R")

    # Save or Return as Bytes
    if as_bytes:
        return pdf.output(dest="S").encode("latin1")  # Return as raw bytes

    pdf.output("invoice.pdf")

File: web-react-app/src/Pages/Homepage/Checkout/Steps/pdfGenerator.js
import jsPDF from "jspdf";
import "jspdf-autotable";
import logo from "./logo.png";

export const generateInvoicePDF = ({
  orderId,
  personalInfo,
  deliveryMethod,
  paymentMethod,
  shoppingBag,
  totalPrice,
  isForView = false,
  asBytes = false,
}) => {
  const doc = new jsPDF();

  // Header
  doc.addImage(logo, "PNG", 85, 10, 40, 20); // Logo centered at the top
  doc.setFontSize(16);
  doc.text("Melangé Invoice", 105, 40, { align: "center" });
  doc.setFontSize(12);
  doc.text(`Order ID: ${orderId}`, 14, 50);

  // Personal Information
  doc.setFontSize(14);
  doc.text("Personal Information", 14, 60);
  doc.line(14, 62, 196, 62); // Horizontal line
  doc.setFontSize(12);
  doc.text(`Name: ${personalInfo.name || "N/A"}`, 14, 70);
  doc.text(`Surname: ${personalInfo.surname || "N/A"}`, 14, 80);
  doc.text(`Address: ${personalInfo.address || "N/A"}`, 14, 90);
  doc.text(`Email: ${personalInfo.email || "N/A"}`, 14, 100);

  // Delivery Method
  doc.setFontSize(14);
  doc.text("Delivery Method", 14, 110);
  doc.line(14, 112, 196, 112); // Horizontal line
  doc.setFontSize(12);
  const deliveryText =
    deliveryMethod === "special"
      ? "Special Delivery (Extra charges apply)"
      : "Standard Delivery";
  doc.text(deliveryText, 14, 120);

  // Payment Information
  doc.setFontSize(14);
  doc.text("Payment Information", 14, 130);
  doc.line(14, 132, 196, 132); // Horizontal line
  doc.setFontSize(12);
  if (paymentMethod?.type === "credit") {
    doc.text(`Card Holder: ${paymentMethod.cardHolderName || "N/A"}`, 14, 140);
    doc.text(
      `Card Number: **** **** **** ${paymentMethod.cardNumber.slice(-4)}`,
      14,
      150
    );
  } else {
    doc.text("No Payment Information Provided", 14, 140);
  }

  // Shopping Bag (Table)
  const tableRows = shoppingBag.map((item) => [
    item.product_name,
    item.color_name,
    item.size_name,
    item.quantity,
    `$${item.price.toFixed(2)}`,
  ]);

  doc.autoTable({
    head: [["Product Name", "Color", "Size", "Quantity", "Price"]],
    body: tableRows,
    startY: 160,
    headStyles: { fillColor: [41, 128, 185], textColor: [255, 255, 255] },
    styles: { fontSize: 10 },
  });

  // Total Price
  const finalY = doc.autoTable.previous.finalY + 10;
  doc.setFontSize(14);
  doc.text(`Total Price: $${totalPrice.toFixed(2)}`, 14, finalY);

  // Output
  if (asBytes) {
    return doc.output("arraybuffer");
  } else if (isForView) {
    const blob = doc.output("blob");
    const url = URL.createObjectURL(blob);
    window.open(url, "_blank");
  } else {
    doc.save(`Invoice_${orderId}.pdf`);
  }
};
